helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java

```diff
@@ -401,11 +401,17 @@ private void refreshBaseline(ResourceControllerDataProvider clusterData,
 
     // Asynchronously report baseline divergence metric before persisting to metadata store,
     // just in case if persisting fails, we still have the metric.
+    // To avoid changes of the new assignment and make it safe when being used to measure baseline
+    // divergence, use a deep copy of the new assignment.
+    Map<String, ResourceAssignment> newAssignmentCopy = new HashMap<>();
+    for (Map.Entry<String, ResourceAssignment> entry : newAssignment.entrySet()) {
+      newAssignmentCopy.put(entry.getKey(), new ResourceAssignment(entry.getValue().getRecord()));
```

Comment:
Luckily we have
```
new ResourceAssignment(ZNRecord)
```
which is creating a deep copy of a resource assignment.

Reply:
ZNRecord has mapFields which is a nested map: Map<String, Map<String, String>>. Caling putAll doesn't do deep copy. Still have to do deep copy manually.